File: Python/Letter Combinations of a Phone Number.py
class Solution:
    # @return a list of strings, [s1, s2]
	def letterCombinations(self, digits):
		trans = [""," ","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"]
		set = []
		seq = ""
		self.generator(trans,digits,0,seq,set)
		return set
	def generator(self,trans,digits,deep,result,set):
		if deep == len(digits):
			newS = result+""
			set.append(newS)
			return
		currentDigit = digits[deep]
		for i in range(0,len(trans[int(currentDigit)])):
			result = result + (trans[int(currentDigit)][i])
			self.generator(trans,digits,deep+1,result,set)
			result = result[0:len(result)-1]

File: Java/Letter Combinations of a Phone Number.java
class Letter Combinations of a Phone Number{
	public List<String> letterCombinations(String digits) {
		String[] trans = {""," ","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        ArrayList<String> res = new ArrayList<String>();
        combinations(digits,0,"",trans,res);
        return res;
	}
	//dfs
	private void combinations(String digits,int level, String curr, String[] trans, ArrayList<String> res){
		if(level == digits.length()){
			res.add(curr);
			return;
		}
		int num = (int)(digits.charAt(level) - '0');
		for(int i=0;i<trans[num].length();i++){
			combinations(digits,level+1,curr+tras[num].charAt(i),res);
		}
	}
}